qiskit_experiments/framework/experiment_data.py

```diff
@@ -2648,24 +2653,48 @@ def __str__(self):
         ret += f"\nArtifacts: {len(self._artifacts)}"
         return ret
 
-    def add_artifacts(self, artifacts: ArtifactData | list[ArtifactData], overwrite: bool = False):
+    @deprecate_arg(
+        name="overwrite",
+        new_alias="overwrite_id",
+        since="0.7",
+        package_name="qiskit-experiments",
+        pending=True,
+    )
+    def add_artifacts(
+        self,
+        artifacts: ArtifactData | list[ArtifactData],
+        overwrite_id: bool = False,
+        overwrite_name: bool = False,
+    ):
         """Add artifacts to experiment. The artifact ID must be unique.
 
+        TODO: update argument list to have only `overwrite` with multiple options
+
         Args:
             artifacts: Artifact or list of artifacts to be added.
-            overwrite: Whether to overwrite the existing artifact.
+            overwrite_id: Whether to overwrite the existing artifact by ID.
+            overwrite_name: Whether to overwrite the existing artifact by name.
         """
+
         if isinstance(artifacts, ArtifactData):
             artifacts = [artifacts]
 
         for artifact in artifacts:
-            if artifact.artifact_id in self._artifacts and not overwrite:
+            if artifact.artifact_id in self._artifacts and not overwrite_id:
                 raise ValueError(
-                    "An artifact with id {artifact.id} already exists."
-                    "Set overwrite to True if you want to overwrite the existing"
+                    f"An artifact with id {artifact.artifact_id} already exists. "
+                    "Set overwrite_id to True if you want to overwrite the existing "
                     "artifact."
                 )
-            self._artifacts[artifact.artifact_id] = artifact
+            if overwrite_name:
+                keys = self._find_artifact_keys(artifact.name)
+                if keys:
+                    for key in keys:
+                        self._artifacts[key] = artifact
+                else:
+                    self._artifacts[artifact.artifact_id] = artifact
+            else:
+                self._artifacts[artifact.artifact_id] = artifact
```

Comment:
The logic here is somewhat strange and unintuitive, I don't know if it's documented anywhere, and it's not clear where it comes from. From the code I see that multiple, say five calls to `add_artifacts` with `overwrite_name=False` will create a collection of five different artifacts having the same name. Until the sixth call sets `overweite_name` to `True`, in this case all the five existing artifacts with that name will remain in the collection, but will be overridden by the new artifact. However a sixth artifact with `artifact_id` will not be added.